torch_audiomentations/augmentations/pitch_shift.py

```diff
@@ -0,0 +1,72 @@
+from random import choices
+import torch
+import torch.nn.functional as F
+
+from ..core.transforms_interface import BaseWaveformTransform
+from torch_pitch_shift import *
+
+
+class PitchShift(BaseWaveformTransform):
+    """
+    Pitch-shift the sound up or down without changing the tempo
+    """
+
+    supports_multichannel = True
+    requires_sample_rate = True
+
+    def __init__(
+        self,
+        sample_rate: int,
+        min_transpose_ratio=0.5,
+        max_transpose_ratio=2,
+        mode: str = "per_example",
+        p: float = 0.5,
+        p_mode: str = None,
+    ):
+        """
+        :param min_transpose_ratio: Minimum pitch shift transposition ratio (default 0.5 --> -1 octaves)
+        :param max_transpose_ratio: Maximum pitch shift transposition ratio (default 2 --> +1 octaves)
+        :param mode:
+        :param p:
+        :param p_mode:
+        :param sample_rate:
+        """
+        super().__init__(mode, p, p_mode, sample_rate)
+
+        self._min_transpose_ratio = min_transpose_ratio
+        self._max_transpose_ratio = max_transpose_ratio
+        if self._min_transpose_ratio > self._max_transpose_ratio:
+            raise ValueError("max_transpose_ratio must be > min_transpose_ratio")
+        self._pitch_shift = PitchShifter()
+        self._sample_rate = sample_rate
+        self._fast_shifts = self.fast_shifts = get_fast_shifts(
+            sample_rate,
+            lambda x: x >= min_transpose_ratio and x <= max_transpose_ratio and x != 1,
+        )
+
+    def randomize_parameters(
+        self, selected_samples: torch.Tensor, sample_rate: int = None
+    ):
+        """
+        :params selected_samples: (batch_size, num_channels, num_samples)
+        """
+        batch_size, _, num_samples = selected_samples.shape
+
+        self.transform_parameters["transpositions"] = choices(
+            self._fast_shifts, k=batch_size
+        )
+
+    def apply_transform(self, selected_samples: torch.Tensor, sample_rate: int = None):
+        batch_size, num_channels, num_samples = selected_samples.shape
+
+        if sample_rate is None:
+            sample_rate = self._sample_rate
+
+        for i in range(batch_size):
```

Comment:
Is there a way to make the computation "batched", so we can get by without this for loop?

Reply:
The shift `__call__` looks like this:

```python
    def __call__(
        self, input: torch.Tensor, shift: Union[float, Fraction], sample_rate: int
    ) -> torch.Tensor:
        """
        Shift the pitch of a waveform by a given amount.

        Parameters
        ----------
        input: torch.Tensor [shape=(channels, samples)]
            Input audio clip of shape (channels, samples)
        shift: float OR Fraction
            `float`: Amount to pitch-shift in # of bins. (1 bin == 1 semitone if `bins_per_octave` == 12)
            `Fraction`: A `fractions.Fraction` object indicating the shift ratio. Usually an element in `get_fast_shifts()`.
        sample_rate: int
            The sample rate of the input audio clip.

        sample_rate: int

        Returns
        -------
        output: torch.Tensor [shape=(channels, samples)]
            The pitch-shifted audio clip
        """
        if not isinstance(shift, Fraction):
            shift = 2.0 ** (float(shift) / self._bins_per_octave)
        resampler = T.Resample(sample_rate, int(sample_rate / shift)).to(input.device)
        output = input
        output = resampler(output)
        output = torch.stft(output, self._n_fft)[None, ...]
        stretcher = T.TimeStretch(
            fixed_rate=float(1 / shift), n_freq=output.shape[2]
        ).to(input.device)
        output = stretcher(output)
        output = torch.istft(output[0], self._n_fft)
        del resampler, stretcher
        if output.shape[1] >= input.shape[1]:
            return output[:, : (input.shape[1])]
        return pad(output, pad=(0, input.shape[1] - output.shape[1], 0, 0))
```

From what I know, there isn't a way to apply variable TimeStretch or Resample operations at variable on a batch. I guess we can add a mode to apply the same random augmentation to each batch so they can be done in parallel?

Reply:
To clarify, if we do add a mode that allows for all samples in one batch to be randomly shifted by the same factor, it would basically be the same as creating one track with a large number of channels. That could definitely work

Reply:
Ok, I see. We could consider it future work then.